### PR TITLE
Rewrite disk redirection using WinPR

### PR DESCRIPTION
**Pull request by DavBfr**
Goal: Handle properly Unicode characters like '表' (U+8868) or '𤭢' (U+24B62 UTF-16 surrogate pairs) with disk redirection.
- Add some missing functions to WinPR
- Rewrite the disk redirection channel using the WinPR *W functions

### CODE REVIEW

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by bmiklautz** [comment]
@freerdp-bot test

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@DavBfr Your pull looks quite nice. Except for the comment from @hardening and the mixed declarations failing VS2010 build. (the latter you can spot by adding ```-DCMAKE_C_FLAGS="-std=gnu90 -pedantic"``` and grep for ```ISO C90 forbids mixed declarations and code```

**Comment by bmiklautz** [comment]
@freerdp-bot test

**Comment by bmiklautz** [comment]
@freerdp-bot test

File: channels/drive/client/drive_file.c [hunk 1]
```diff
@@ -595,15 +573,12 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
 				return FALSE;
 
-			hFd = CreateFileA(file->fullpath, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING,
-			                  FILE_ATTRIBUTE_NORMAL, NULL);
-
-			if (hFd == INVALID_HANDLE_VALUE)
+			if (file->file_handle == INVALID_HANDLE_VALUE)
 			{
-				WLog_ERR(TAG, "Unable to create file %s", file->fullpath);
+				WLog_ERR(TAG, "Unable to set file time %s (%d)", file->fullpath, GetLastError());
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: Should use PRId32, ```GetLastError``` returns a DWORD

File: channels/drive/client/drive_file.c [hunk 2]
```diff
@@ -635,14 +610,15 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
 			}
 
-			if (!SetFileTime(hFd, pftCreationTime, pftLastAccessTime, pftLastWriteTime))
+			DEBUG_WSTR("SetFileTime %s", file->fullpath);
+
+			if (!SetFileTime(file->file_handle, pftCreationTime, pftLastAccessTime, pftLastWriteTime))
 			{
-				WLog_ERR(TAG, "Unable to set file time on %s", file->fullpath);
-				CloseHandle(hFd);
+				WLog_ERR(TAG, "Unable to set file time %s to %d", file->fullpath);
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: Missing argument

File: channels/drive/client/drive_file.c [hunk 3]
```diff
@@ -651,86 +627,98 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
 			/* http://msdn.microsoft.com/en-us/library/cc232076.aspx */
 			Stream_Read_INT64(input, size);
-			hFd = CreateFileA(file->fullpath, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING,
-			                  FILE_ATTRIBUTE_NORMAL, NULL);
 
-			if (hFd == INVALID_HANDLE_VALUE)
+			if (file->file_handle == INVALID_HANDLE_VALUE)
 			{
-				WLog_ERR(TAG, "Unable to truncate %s to %"PRId64"", file->fullpath, size);
+				WLog_ERR(TAG, "Unable to truncate %s to %"PRId64" (%d)", file->fullpath, size, GetLastError());
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: Again PRId32

File: channels/drive/client/drive_file.c [hunk 4]
```diff
@@ -651,86 +627,98 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
-			liSize.QuadPart = size;
+			liSize.QuadPart = size & 0xFFFFFFFF;
 
-			if (SetFilePointer(hFd, liSize.LowPart, &liSize.HighPart, FILE_BEGIN) == 0)
+			if (SetFilePointer(file->file_handle, liSize.LowPart, &liSize.HighPart,
+			                   FILE_BEGIN) == INVALID_SET_FILE_POINTER)
 			{
-				WLog_ERR(TAG, "Unable to truncate %s to %"PRId64"", file->fullpath, size);
-				CloseHandle(hFd);
+				WLog_ERR(TAG, "Unable to truncate %s to %d (%d)", file->fullpath, size, GetLastError());
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: Again PRId32

File: channels/drive/client/drive_file.c [hunk 5]
```diff
@@ -651,86 +627,98 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
+				return FALSE;
+			}
+
+			DEBUG_WSTR("Truncate %s", file->fullpath);
+
+			if (SetEndOfFile(file->file_handle) == 0)
+			{
+				WLog_ERR(TAG, "Unable to truncate %s to %d (%d)", file->fullpath, size, GetLastError());
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: Again PRId32

File: winpr/libwinpr/file/file.c [hunk 6]
```diff
@@ -97,10 +105,12 @@ static BOOL FileSetEndOfFile(HANDLE hFile)
 		return FALSE;
 
 	size = ftell(pFile->fp);
+	errno = 0;
 	if (ftruncate(fileno(pFile->fp), size) < 0)
```
Comment (hardening) [inline winpr/libwinpr/file/file.c]: it doesn't hurt but it is not needed.

File: channels/drive/client/drive_main.c [hunk 7]
```diff
@@ -68,37 +72,59 @@ struct _DRIVE_DEVICE
...
+			rc = STATUS_NOT_A_DIRECTORY;
+			break;
+
+		case ERROR_PATH_NOT_FOUND:
+			rc = STATUS_OBJECT_PATH_NOT_FOUND;
 			break;
 
 		default:
 			rc = STATUS_UNSUCCESSFUL;
+			WLog_ERR(TAG, "Error code not found: %d", fs_errno);
```
Comment (akallabeth) [inline channels/drive/client/drive_main.c]: Again PRId32

File: winpr/libwinpr/crt/string.c [hunk 8]
```diff
@@ -129,6 +129,20 @@ WCHAR* _wcschr(const WCHAR* str, WCHAR c)
...
 }
 
+/* _wcsrchr -> wcsrchr */
+
+WCHAR* _wcsrchr(const WCHAR* str, WCHAR c)
+{
+	WCHAR *p;
+	WCHAR ch;
+
+	for (p = (WCHAR *) 0; (ch = *str); str++)
```
Comment (akallabeth) [inline winpr/libwinpr/crt/string.c]: Is str=NULL allowed?

File: winpr/libwinpr/file/file.c [hunk 9]
```diff
@@ -161,9 +171,10 @@ static BOOL FileRead(PVOID Object, LPVOID lpBuffer, DWORD nNumberOfBytesToRead,
 		return FALSE;
 
 	file = (WINPR_FILE *)Object;
-	io_status = fread(lpBuffer, nNumberOfBytesToRead, 1, file->fp);
+	errno = 0;
```
Comment (hardening) [inline winpr/libwinpr/file/file.c]: taken from man fread:
```
fread() does not distinguish between end-of-file and error, and callers
       must use feof(3) and ferror(3) to determine which occurred.
```
so the errno tricks aren't accurate

File: winpr/libwinpr/path/shell.c [hunk 10]
```diff
@@ -505,9 +507,54 @@ BOOL PathFileExistsA(LPCSTR pszPath)
...
+	return ret;
 }
 
+BOOL PathIsDirectoryEmptyA(LPCSTR pszPath)
+{
+	struct dirent *dp;
+	int empty = 1;
+
+	DIR *dir = opendir(pszPath);
+	if (dir == NULL) //Not a directory or doesn't exist
```
Comment (akallabeth) [inline winpr/libwinpr/path/shell.c]: c++ comment

File: winpr/libwinpr/file/file.c [hunk 11]
```diff
@@ -199,11 +212,15 @@ static BOOL FileWrite(PVOID Object, LPCVOID lpBuffer, DWORD nNumberOfBytesToWrit
 
 	file = (WINPR_FILE *)Object;
 
-	io_status = fwrite(lpBuffer, nNumberOfBytesToWrite, 1, file->fp);
-	if (io_status != 1)
+	errno = 0;
```
Comment (hardening) [inline winpr/libwinpr/file/file.c]: The same here for errno tricks

**Comment by hardening** [comment]
As a general comment, I'm starting to think that all File operations should be done on the file descriptor instead of using f[read|write] operations, so that we can correctly map errno to a windows error. 
One would even suggest that it would be a good idea that FILE winpr object would not have use FILE * but just the raw file descriptor...

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot be nice and test this one trough!

File: winpr/libwinpr/file/generic.c [hunk 12]
```diff
@@ -774,15 +913,92 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
+
+	if (FindNextFileA(hFindFile, fd))
+	{
+		CopyMemory(lpFindFileData, fd, 352);
+
+		WCHAR* unicodeFileName = NULL;
+		int length = ConvertToUnicode(CP_UTF8, 0, fd->cFileName, -1, &unicodeFileName, 0) * 2;
+		if (length == 0)
+		{
+			SetLastError(ERROR_NOT_ENOUGH_MEMORY);
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Memory leak ```fd```

File: winpr/libwinpr/file/generic.c [hunk 13]
```diff
@@ -774,15 +913,92 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
+				lpFindFileData->dwFileAttributes = FILE_ATTRIBUTE_ARCHIVE;
+
+			if (pFileSearch->pDirent->d_name[0] == '.' && namelen != 1 &&
+				 (pFileSearch->pDirent->d_name[1] != '.' && namelen != 2))
+				lpFindFileData->dwFileAttributes |= FILE_ATTRIBUTE_HIDDEN;
+
+			if (! (fileStat.st_mode & S_IWUSR))
+				lpFindFileData->dwFileAttributes |= FILE_ATTRIBUTE_READONLY;
+
+			UINT64 ft;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Mixed declaration

File: winpr/libwinpr/file/generic.c [hunk 14]
```diff
@@ -721,45 +834,71 @@ HANDLE FindFirstFileA(LPCSTR lpFileName, LPWIN32_FIND_DATAA lpFindFileData)
...
+		return INVALID_HANDLE_VALUE;
+	}
+
+	if (ConvertFromUnicode(CP_UTF8, 0, lpFileName, -1, &utfFileName, 0, NULL, NULL) <= 0)
+	{
+		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+		return INVALID_HANDLE_VALUE;
+	}
+
+	HANDLE h = FindFirstFileA(utfFileName, fd);
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Mixed declaration

File: winpr/libwinpr/file/generic.c [hunk 15]
```diff
@@ -721,45 +834,71 @@ HANDLE FindFirstFileA(LPCSTR lpFileName, LPWIN32_FIND_DATAA lpFindFileData)
...
+	free(utfFileName);
+
+	if (h != INVALID_HANDLE_VALUE)
+	{
+		CopyMemory(lpFindFileData, fd, 352);
+
+		WCHAR* unicodeFileName = NULL;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Mixed declaration

File: winpr/libwinpr/file/generic.c [hunk 16]
```diff
@@ -721,45 +834,71 @@ HANDLE FindFirstFileA(LPCSTR lpFileName, LPWIN32_FIND_DATAA lpFindFileData)
...
+		int length = ConvertToUnicode(CP_UTF8, 0, fd->cFileName, -1, &unicodeFileName, 0) * 2;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Mixed declaration

File: winpr/libwinpr/file/generic.c [hunk 17]
```diff
@@ -774,15 +913,92 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
+	{
+		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+		return INVALID_HANDLE_VALUE;
+	}
+
+	if (FindNextFileA(hFindFile, fd))
+	{
+		CopyMemory(lpFindFileData, fd, 352);
+
+		WCHAR* unicodeFileName = NULL;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Mixed declaration

File: channels/drive/client/drive_file.c [hunk 18]
```diff
@@ -651,86 +627,98 @@ BOOL drive_file_set_information(DRIVE_FILE* file, UINT32 FsInformationClass, UIN
...
 			break;
 
 		case FileDispositionInformation:
 
 			/* http://msdn.microsoft.com/en-us/library/cc232098.aspx */
 			/* http://msdn.microsoft.com/en-us/library/cc241371.aspx */
-			if (file->is_dir && !dir_empty(file->fullpath))
-				break;
+			if (file->is_dir && !PathIsDirectoryEmptyW(file->fullpath))
+				break; // TODO: SetLastError ???
```
Comment (akallabeth) [inline channels/drive/client/drive_file.c]: c++ comment

File: winpr/libwinpr/file/file.c [hunk 19]
```diff
@@ -695,6 +779,43 @@ BOOL SetStdHandleEx(DWORD dwStdHandle, HANDLE hNewHandle, HANDLE* phOldHandle)
 	return FALSE;
 }
 
+BOOL GetDiskFreeSpaceA(LPCSTR lpRootPathName, LPDWORD lpSectorsPerCluster,
+											 LPDWORD lpBytesPerSector, LPDWORD lpNumberOfFreeClusters, LPDWORD lpTotalNumberOfClusters)
+{
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__)
+#define STATVFS statvfs
```
Comment (akallabeth) [inline winpr/libwinpr/file/file.c]: Kind of redundant, same as else

File: winpr/libwinpr/file/generic.c [hunk 20]
```diff
@@ -774,15 +913,92 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
+			if (fullpath == NULL) {
+				return INVALID_HANDLE_VALUE;
+			}
+			memcpy(fullpath, pFileSearch->lpPath, pathlen);
+			fullpath[pathlen] = '/';
+			memcpy(fullpath + pathlen + 1, pFileSearch->pDirent->d_name, namelen);
+			fullpath[pathlen+namelen+1] = 0;
+
+			if (lstat(fullpath, &fileStat) != 0)
+				return INVALID_HANDLE_VALUE;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: Memory leak ```fullpath```

**Comment by akallabeth** [comment]
@DavBfr Could you be so kind to fix the issues found in the latest reviews? Would be nice to merge this one finally.

File: winpr/libwinpr/file/generic.c (hunk 12 above)
Comment (DavBfr) [inline winpr/libwinpr/file/generic.c]: Freed the line just after

Comment (DavBfr) [inline winpr/libwinpr/file/generic.c]: Freed the line just after

**Comment by akallabeth** [comment]
@freerdp-bot test

File: winpr/libwinpr/file/generic.c [hunk 21]
```diff
@@ -774,15 +919,101 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
 BOOL FindNextFileW(HANDLE hFindFile, LPWIN32_FIND_DATAW lpFindFileData)
 {
+	WCHAR* unicodeFileName;
+	int length;
+
+	LPWIN32_FIND_DATAA fd = (LPWIN32_FIND_DATAA)malloc(sizeof(WIN32_FIND_DATAA));
+	if (!fd)
+	{
+		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+		return INVALID_HANDLE_VALUE;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: @DavBfr Wrong return value

File: winpr/libwinpr/file/generic.c [hunk 22]
```diff
@@ -774,15 +919,101 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
 		if (FilePatternMatchA(pFileSearch->pDirent->d_name, pFileSearch->lpPattern))
 		{
 			strcpy(lpFindFileData->cFileName, pFileSearch->pDirent->d_name);
+			namelen = strlen(lpFindFileData->cFileName);
+
+			pathlen = strlen(pFileSearch->lpPath);
+			fullpath = (char*)malloc(pathlen + namelen + 2);
+			if (fullpath == NULL)
+			{
+				return INVALID_HANDLE_VALUE;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: @DavBfr Wrong return value

File: winpr/libwinpr/file/generic.c [hunk 23]
```diff
@@ -774,15 +919,101 @@ BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
...
+			}
+			memcpy(fullpath, pFileSearch->lpPath, pathlen);
+			fullpath[pathlen] = '/';
+			memcpy(fullpath + pathlen + 1, pFileSearch->pDirent->d_name, namelen);
+			fullpath[pathlen+namelen+1] = 0;
+
+			if (lstat(fullpath, &fileStat) != 0)
+			{
+				free(fullpath);
+				return INVALID_HANDLE_VALUE;
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: @DavBfr Wrong return value

File: winpr/libwinpr/file/generic.c [hunk 24]
```diff
@@ -461,6 +468,112 @@ BOOL FlushFileBuffers(HANDLE hFile)
...
 }
 
+BOOL WINAPI GetFileAttributesExA(LPCSTR lpFileName, GET_FILEEX_INFO_LEVELS fInfoLevelId,
+																 LPVOID lpFileInformation)
+{
+	LPWIN32_FILE_ATTRIBUTE_DATA fd = lpFileInformation;
+	WIN32_FIND_DATAA findFileData;
+	HANDLE hFind;
+
+	if (!(hFind = FindFirstFileA(lpFileName, &findFileData) == INVALID_HANDLE_VALUE))
```
Comment (DavBfr) [inline winpr/libwinpr/file/generic.c]: Oh yes, it's the same right?

File: winpr/libwinpr/file/generic.c (hunk 21 above)
Comment (DavBfr) [inline winpr/libwinpr/file/generic.c]: Fixed.

**Comment by akallabeth** [comment]
@freerdp-bot test

File: winpr/libwinpr/file/generic.c (hunk 24 above)
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: @DavBfr Not really, only if the inversion of hFind equals ```INVALID_HANDLE_VALUE```. That really depends on the definition. It also produces compiler warnings with C11

File: winpr/libwinpr/file/generic.c [hunk 25]
```diff
@@ -461,6 +468,112 @@ BOOL FlushFileBuffers(HANDLE hFile)
 	return FALSE;
 }
 
+BOOL WINAPI GetFileAttributesExA(LPCSTR lpFileName, GET_FILEEX_INFO_LEVELS fInfoLevelId,
+																 LPVOID lpFileInformation)
```
Comment (hardening) [inline winpr/libwinpr/file/generic.c]: indentation looks strange here, at least in the github view...

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@hardening Do you think it is ready now?

**Comment by akallabeth** [comment]
@freerdp-bot test

**Pull request merged**